src/java.base/share/native/libjli/java.c

```diff
@@ -534,6 +545,260 @@ invokeInstanceMainWithoutArgs(JNIEnv *env, jclass mainClass) {
+                freeCommandLineOptionValue(logFilePropertyValue);
+                *error = MEMORY_ALLOCATION_ERROR;
+                return NULL;
+            }
+            char *logFile = (char *)JLI_MemAlloc(logFileLength);
+            if (NULL == logFile) {
+                JLI_ReportErrorMessage("Failed to allocate memory for the CRIU log file option.");
+                *error = MEMORY_ALLOCATION_ERROR;
+                return NULL;
+            }
+            snprintf(logFile, logFileLength, "--log-file=%s", logFilePropertyValue);
+            freeCommandLineOptionValue(logFilePropertyValue);
+            return logFile;
+        } else {
+            JLI_ReportErrorMessage("The value of the command line option %s was not found.", logFilePropertyName);
+            *error = OPTION_VALUE_NOT_FOUND;
+            return NULL;
+        }
+    }
+    return NULL;
+}
+
+static int restoreFromCheckpoint(char *checkpointDir, char *logLevel, char *unprivilegedMode, char *logFile) {
+    int argc = 6;
+    char *argv[9] = { NULL };
+    argv[0] = "criu";
+    argv[1] = "restore";
+    argv[2] = "-D";
+    argv[3] = checkpointDir;
+    argv[4] = logLevel;
+    argv[5] = "--shell-job";
+    if (unprivilegedMode) {
+        argv[argc++] = unprivilegedMode;
+    }
+    if (logFile) {
+        argv[argc++] = logFile;
+    }
+    argv[argc] = NULL;
+    if (-1 == execvp(argv[0], argv)) {
+        return -1;
+    }
+    return 0;
+}
+
+static void handleCRaCRestore(int argc, char **argv) {
+    int error = 0;
+    char *checkpointDir = getCheckpointDir(argc, argv, &error);
+    if (OPTION_NOT_FOUND == error) {
+        freeCommandLineOptionValue(checkpointDir);
+        return;
+    }
+    if (error) {
+        JLI_ReportErrorMessage("Failed to get the CRIU checkpoint directory, error=%d.", error);
+        exit(EXIT_FAILURE);
+    }
+    /*
+     * The if block will be invoked by the child process,
+     * and the else block will be invoked by the parent process.
+     */
+    pid_t criuRestorePid = fork();
+    if (0 == criuRestorePid) {
+        char *logLevel = getLogLevel(argc, argv, &error);
+        if (isCommandLineOptionFoundWithError(error)) {
+            JLI_ReportErrorMessage("Failed to get the CRIU log level, error=%d.", error);
+            exit(EXIT_FAILURE);
+        }
+        char *unprivilegedMode = getUnprivilegedMode(argc, argv, &error);
+        if (isCommandLineOptionFoundWithError(error)) {
+            JLI_ReportErrorMessage("Failed to get the CRIU unprivileged mode, error=%d.", error);
+            freeCommandLineOptionValue(logLevel);
+            exit(EXIT_FAILURE);
+        }
+        char *logFile = getLogFile(argc, argv, &error);
+        if (isCommandLineOptionFoundWithError(error)) {
+            JLI_ReportErrorMessage("Failed to get the CRIU log file, error=%d.", error);
+            freeCommandLineOptionValue(logLevel);
+            freeCommandLineOptionValue(unprivilegedMode);
```

Comment:
The duplicated `freeCommandLineOptionValue()` code might be reduced if a cleanup label is added at the end of this method.